Fix the UIOpacity issue when the node is re-added to the scene.

yoki0805 opened the pull request:
issue:
[https://forum.cocos.org/t/topic/159379/745](https://forum.cocos.org/t/topic/159379/745)

`
if (JSB) { render.node._uiProps.localOpacity = render.renderEntity.localOpacity; }
`
'render.node._uiProps.localOpacity' still needs to be updated on native for fix another issue.
[#17503](https://github.com/cocos/cocos-engine/pull/17503)

If possible, it would be great to have this fix synced to v3.8.4 as well.
@minggo 

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

<!-- Note: Makes sure these boxes are checked before submitting your PR - thank you!
- [ ] Your pull request title is using English, it's precise and appropriate.
- [ ] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
- [ ] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
- [ ] Document new code with comments in source code based on API docs
- [ ] Make sure any runtime log information in `log` , `error` or `new Error('')` has been moved into `EngineErrorMap.md` with an ID, and use `logID(id)` or `new Error(getError(id))` instead.
- To official teams:
  - [ ] Check that your PR is following our [guides](https://github.com/cocos/3d-tasks/blob/master/workflows/readme.md)
-->

yoki0805 committed:
fix ui-opacity

fix the render.node._uiProps.localOpacity only update on native.

minggo commented:
@yoki0805 thanks for the PR. As v3.8.4 is releasing, so i can't merge it to v3.8.4.

yoki0805 commented:
> @yoki0805 thanks for the PR. As v3.8.4 is releasing, so i can't merge it to v3.8.4.

When v3.8.4 is released, could this issue be included in the known issues list?

minggo commented:
It may be released tomorrow. I will add it to known list in release note if we have.

qiuguohua commented:
@yoki0805 Sorry I clicked wrong. My suggestion is to add the judgment in _parentChanged.
```
    protected _parentChanged (): void {
        if (!JSB) {
            return;
        }
```
The setEntityLocalOpacityDirtyRecursively function should not work on web platforms!

yoki0805 committed:
Add JSB Check in _parentChanged function

The setEntityLocalOpacityDirtyRecursively function should not work on web platform.

yoki0805 committed:
fix the opacity issue on native

Whenever the node is re-added to the scene, the _parentOpacity value needs to be updated again in the onEnable callback.

yoki0805 committed:
fix ESLint warning

yoki0805 commented:
1. The JSB checking is added.
```
    protected _parentChanged (): void {
        if (!JSB) {
            return;
        }
```

2. Fixed an issue on native platform where a node wasn't affected by its parent's transparency when added to the scene.
 - The node will wait for 2 seconds and then fade out after being added to the scene. (The node is affected by its parent's transparency.)
 - Test demo: [UIOpacity.zip](https://github.com/user-attachments/files/17344124/UIOpacity.zip)

@qiuguohua

qiuguohua commented on cocos/2d/components/ui-opacity.ts:
```diff
@@ -152,7 +154,9 @@ export class UIOpacity extends Component {
                 render.renderEntity.localOpacity = parentOpacity;
             }
-            render.node._uiProps.localOpacity = render.renderEntity.localOpacity;
+            if (JSB) {
```
It should be possible to remove it here? @yoki0805

yoki0805 replied:
```typescript
if (JSB) {
    render.node._uiProps.localOpacity = render.renderEntity.localOpacity;
}
```
If I remove it, the issue [#17503](https://github.com/cocos/cocos-engine/pull/17503) will reoccur on native platform.
![截圖 2024-10-12 上午10 23 38](https://github.com/user-attachments/assets/f79e5efe-31b4-47ce-8627-97a4222cc60e)

yoki0805 replied:
I got what you mean.
UIOpacity.setEntityLocalOpacityDirtyRecursively is only called on native platform.
I will remove the JSB checking here, thank you.

qiuguohua commented on cocos/2d/components/ui-opacity.ts:
```diff
@@ -222,6 +231,10 @@ export class UIOpacity extends Component {
 
     public onEnable (): void {
+        if (this._parentOpacityResetFlag) {
+            this._parentChanged();
```
It is recommended to change it to something like the following
```
        this.node.on(NodeEventType.PARENT_CHANGED, this._parentChanged, this);
        if (this._parentOpacityResetFlag) {
            this._parentChanged();
            this._parentOpacityResetFlag = false;
        } else {
            this.node._uiProps.localOpacity = this._parentOpacity * this._opacity / 255;
            this._setEntityLocalOpacityRecursively(this.node._uiProps.localOpacity);
        }
```

yoki0805 replied:
```typescript
public onEnable (): void {
    this.node.on(NodeEventType.PARENT_CHANGED, this._parentChanged, this);
    if (this._parentOpacityResetFlag) {
        this._parentChanged();
        this._parentOpacityResetFlag = false;
    } else {
        this.node._uiProps.localOpacity = this._parentOpacity * this._opacity / 255;
        this._setEntityLocalOpacityRecursively(this.node._uiProps.localOpacity);
    }
}
```
I changed to that, and I got the different results in test demo.
on web platform:
![截圖 2024-10-12 上午10 35 33](https://github.com/user-attachments/assets/d295409c-ed90-42af-bf10-d30f5ef6bb64)

on native platform:
![截圖 2024-10-12 上午10 36 02](https://github.com/user-attachments/assets/0d2bc833-8fdf-4b13-a1b7-353216e66558)

The result on native platform is right.

qiuguohua replied:
Sorry, should also need to set this.node._uiProps.localOpacity.
```
public onEnable (): void {
    this.node.on(NodeEventType.PARENT_CHANGED, this._parentChanged, this);
    this.node._uiProps.localOpacity = this._parentOpacity * this._opacity / 255;
    if (this._parentOpacityResetFlag) {
        this._parentChanged();
        this._parentOpacityResetFlag = false;
    } else {
        this._setEntityLocalOpacityRecursively(this.node._uiProps.localOpacity);
    }
}
```

yoki0805 replied:
It looks good in test demo, thanks!

yoki0805 committed:
Remove the JSB checking in setEntityLocalOpacityDirtyRecursively function

yoki0805 committed:
Update onEnable function

minggo commented:
@qiuguohua please update release note.
